src/SessionManager.php

```diff
@@ -369,50 +324,33 @@ public function forgetMe(): static
         return $this;
     }
 
-    /**
-     * Set the validator chain to use when validating a session
-     *
-     * In most cases, you should use an instance of {@link ValidatorChain}.
-     */
-    public function setValidatorChain(EventManagerInterface $chain): static
-    {
-        $this->validatorChain = $chain;
-        return $this;
-    }
-
-    /**
-     * Get the validator chain to use when validating a session
-     *
-     * By default, uses an instance of {@link ValidatorChain}.
-     */
-    public function getValidatorChain(): EventManagerInterface
-    {
-        if (null === $this->validatorChain) {
-            $this->setValidatorChain(new ValidatorChain($this->getStorage()));
-            assert($this->validatorChain instanceof EventManagerInterface);
-        }
-        return $this->validatorChain;
-    }
-
     /**
      * Is this session valid?
      *
      * Notifies the Validator Chain until either all validators have returned
      * true or one has failed.
      */
-    public function isValid(): bool
+    public function isValid(): void
     {
-        $validator = $this->getValidatorChain();
-        $event     = new Event();
-        $event->setName('session.validate');
-        $event->setTarget($this);
-        $event->setParams($this);
+        /** @var array<string, mixed> $storage */
+        $storage = $this->getStorage()->getMetadata();
 
-        $falseResult = static fn($test): bool => false === $test;
+        if (isset($storage['environment'])) {
+            assert(is_string($storage['environment']));
+            /** @var EnvironmentInterface $initialEnvironment */
+            $initialEnvironment = unserialize($storage['environment']);
+        } else {
+            $initialEnvironment = $this->environmentFactory->getEnvironment();
+            $this->getStorage()->setMetadata('environment', serialize($initialEnvironment));
+        }
 
-        $responses = $validator->triggerEventUntil($falseResult, $event);
+        foreach ($this->validators as $validator) {
+            if (! $validator instanceof ValidatorInterface) {
+                continue;
+            }
 
-        return ! $responses->stopped();
+            $validator->validate($initialEnvironment, $this->environmentFactory->getEnvironment());
```

Comment:
Possibly fetch the current environment in a variable outside of the loop. It's not doing any heavy lifting here, but other implementations might.